# PART D

In [29]:
%pip install -q pandas numpy scikit-learn nltk tokenizers

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.0.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [30]:
import pandas as pd
import gzip
import json
import numpy as np

In [31]:
from pathlib import Path

data_filename = 'c4-train.00000-of-01024-30K.json.gz'
data_candidates = [Path('../data') / data_filename, Path('data') / data_filename]
data_path = next((path for path in data_candidates if path.exists()), None)
if data_path is None:
    raise FileNotFoundError('Không tìm thấy corpus. Đặt file trong ../data hoặc data so với working directory.')

documents = []
with gzip.open(data_path, 'rt', encoding='utf-8') as f:
    for line in f:
        if line.strip():
            documents.append(json.loads(line))

In [32]:
print(documents[0])  # Print the first document to verify the content

{'text': 'Beginners BBQ Class Taking Place in Missoula!\nDo you want to get better at making delicious BBQ? You will have the opportunity, put this on your calendar now. Thursday, September 22nd join World Class BBQ Champion, Tony Balay from Lonestar Smoke Rangers. He will be teaching a beginner level class for everyone who wants to get better with their culinary skills.\nHe will teach you everything you need to know to compete in a KCBS BBQ competition, including techniques, recipes, timelines, meat selection and trimming, plus smoker and fire information.\nThe cost to be in the class is $35 per person, and for spectators it is free. Included in the cost will be either a t-shirt or apron and you will be tasting samples of each meat that is prepared.', 'timestamp': '2019-04-25T12:57:54Z', 'url': 'https://klyq.com/beginners-bbq-class-taking-place-in-missoula/'}


In [33]:
print(len(documents))  # Print the number of documents loaded

30000


In [34]:
df = pd.DataFrame(documents)
del documents
df

,text,timestamp,url
0,Beginners BBQ Class Taking Place in Missoula!\...,2019-04-25T12:57:54Z,https://klyq.com/beginners-bbq-class-taking-pl...
1,Discussion in 'Mac OS X Lion (10.7)' started b...,2019-04-21T10:07:13Z,https://forums.macrumors.com/threads/restore-f...
2,Foil plaid lycra and spandex shortall with met...,2019-04-25T10:40:23Z,https://awishcometrue.com/Catalogs/Clearance/T...
3,How many backlinks per day for new site?\nDisc...,2019-04-21T12:46:19Z,https://www.blackhatworld.com/seo/how-many-bac...
4,The Denver Board of Education opened the 2017-...,2019-04-20T14:33:21Z,http://bond.dpsk12.org/category/news/
...,...,...,...
29995,HatchViewSPC turns your USB microscope into a ...,2019-04-23T04:46:14Z,https://digitalmetrology.myshopify.com/collect...
29996,TechCrackHouse / Investing Made Easy!\nSteve T...,2019-04-25T16:54:28Z,https://www.mansworldmedia.com/category/steve-...
29997,"Coyote Music: You've lived in Boston, so you k...",2019-04-20T10:31:16Z,https://www.coyotemusic.com/publications/detai...
29998,"Pets deserve food that is tasty, nutritious, a...",2019-04-25T22:36:46Z,http://flintriver.com/


In [35]:
df1 = df['text']  # Extract the 'text' column as a list of strings
df1

0        Beginners BBQ Class Taking Place in Missoula!\...
1        Discussion in 'Mac OS X Lion (10.7)' started b...
2        Foil plaid lycra and spandex shortall with met...
3        How many backlinks per day for new site?\nDisc...
4        The Denver Board of Education opened the 2017-...
                               ...                        
29995    HatchViewSPC turns your USB microscope into a ...
29996    TechCrackHouse / Investing Made Easy!\nSteve T...
29997    Coyote Music: You've lived in Boston, so you k...
29998    Pets deserve food that is tasty, nutritious, a...
29999    In a trial case and when cross examining a wit...
Name: text, Length: 30000, dtype: object

In [36]:
print("Number of documents:", len(df1))

Number of documents: 30000


In [37]:
df1_lowercase = df1.str.lower()  # Convert all text to lowercase
df1_lowercase

0        beginners bbq class taking place in missoula!\...
1        discussion in 'mac os x lion (10.7)' started b...
2        foil plaid lycra and spandex shortall with met...
3        how many backlinks per day for new site?\ndisc...
4        the denver board of education opened the 2017-...
                               ...                        
29995    hatchviewspc turns your usb microscope into a ...
29996    techcrackhouse / investing made easy!\nsteve t...
29997    coyote music: you've lived in boston, so you k...
29998    pets deserve food that is tasty, nutritious, a...
29999    in a trial case and when cross examining a wit...
Name: text, Length: 30000, dtype: object

In [38]:
import re

def tokenize_normalized(text):
    text = text.lower()
    text = re.sub(r"[^\w\s]", " ", text)
    tokens = re.findall(r"\b\w+\b", text)
    return tokens

In [39]:
df1_tokens = df1.apply(tokenize_normalized) 
df1_tokens

0        [beginners, bbq, class, taking, place, in, mis...
1        [discussion, in, mac, os, x, lion, 10, 7, star...
2        [foil, plaid, lycra, and, spandex, shortall, w...
3        [how, many, backlinks, per, day, for, new, sit...
4        [the, denver, board, of, education, opened, th...
                               ...                        
29995    [hatchviewspc, turns, your, usb, microscope, i...
29996    [techcrackhouse, investing, made, easy, steve,...
29997    [coyote, music, you, ve, lived, in, boston, so...
29998    [pets, deserve, food, that, is, tasty, nutriti...
29999    [in, a, trial, case, and, when, cross, examini...
Name: text, Length: 30000, dtype: object

In [40]:
df1_tokens[1321]

['so',
 'much',
 'for',
 'global',
 'warming',
 'we',
 'really',
 'need',
 'to',
 'prepare',
 'for',
 'what',
 'happens',
 'if',
 'we',
 'continue',
 'down',
 'this',
 'path',
 'of',
 'colder',
 'winters',
 'the',
 'ground',
 'freezes',
 'and',
 'it',
 'can',
 'freeze',
 'even',
 'down',
 'two',
 'feet',
 'that',
 'means',
 'good',
 'luck',
 'on',
 'growing',
 'anything',
 'it',
 'has',
 'been',
 'so',
 'cold',
 'that',
 'even',
 'the',
 'rails',
 'for',
 'trains',
 'broke',
 'in',
 'three',
 'different',
 'places',
 'between',
 'baltimore',
 'and',
 'washington',
 'on',
 'thursday',
 'so',
 'even',
 'trains',
 'were',
 'delayed',
 'or',
 'canceled',
 'as',
 'amtrak',
 'did',
 'for',
 'trains',
 'passing',
 'through',
 'chicago',
 'it',
 'has',
 'been',
 'so',
 'cold',
 'the',
 'tracks',
 'freeze',
 'and',
 'the',
 'only',
 'way',
 'to',
 'switch',
 'tracks',
 'is',
 'to',
 'set',
 'them',
 'on',
 'fire',
 'of',
 'course',
 'the',
 'tracks',
 'themselves',
 'were',
 'not',
 'truly',
 '

In [43]:
from sklearn.feature_extraction.text import CountVectorizer

count_vectorizer = CountVectorizer(
    tokenizer=str.split,
    token_pattern=None,
    lowercase=False
)

X_count = count_vectorizer.fit_transform(
    df1_tokens.apply(lambda tokens: ' '.join(tokens))
)

In [ ]:
print(X_count)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 5104560 stored elements and shape (30000, 193837)>
  Coords	Values
  (0, 25936)	1
  (0, 25322)	4
  (0, 40418)	4
  (0, 165888)	1
  (0, 132084)	1
  (0, 86528)	4
  (0, 112594)	1
  (0, 54443)	1
  (0, 187807)	5
  (0, 182095)	1
  (0, 170478)	5
  (0, 73073)	2
  (0, 27134)	2
  (0, 21279)	1
  (0, 105923)	1
  (0, 50491)	1
  (0, 184444)	5
  (0, 79366)	1
  (0, 168465)	4
  (0, 124246)	1
  (0, 137988)	1
  (0, 169097)	1
  (0, 123639)	1
  (0, 187859)	1
  (0, 34361)	1
  :	:
  (29999, 149645)	1
  (29999, 85116)	1
  (29999, 100283)	1
  (29999, 173196)	2
  (29999, 143431)	1
  (29999, 128644)	1
  (29999, 122271)	2
  (29999, 185050)	4
  (29999, 185052)	1
  (29999, 63215)	2
  (29999, 64763)	2
  (29999, 187838)	1
  (29999, 46624)	3
  (29999, 99854)	3
  (29999, 21673)	1
  (29999, 156912)	1
  (29999, 51785)	1
  (29999, 129642)	2
  (29999, 88386)	1
  (29999, 47214)	1
  (29999, 63207)	1
  (29999, 183008)	1
  (29999, 89926)	1
  (29999, 138847)	2
  (29999,

In [ ]:
print(X_count.shape)
print(X_count[:3].toarray().max()) # Print the maximum count value

(30000, 193837)
19


In [ ]:
print("Matrix shape:", X_count.shape)
print("Vocabulary size:", len(count_vectorizer.vocabulary_))

Matrix shape: (30000, 193837)
Vocabulary size: 193837


In [ ]:
vocabulary = count_vectorizer.get_feature_names_out()
vocabulary

array(['0', '00', '000', ..., 'floors', 'fluid', '𐌼𐌿𐌽𐌳𐍃'],
      shape=(193837,), dtype=object)

In [ ]:
#Tính độ thưa
nnz = X_count.nnz  # Số zero entries trong ma trận sparse
print(1 - nnz / (X_count.shape[0] * X_count.shape[1]))

0.9991221902939067


In [ ]:
import numpy as np

# Df
document_frequency = np.asarray((X_count > 0).sum(axis=0)).ravel()
top_df_indices = np.argsort(document_frequency)[::-1][:20]

top_df = pd.DataFrame({
    'term': vocabulary[top_df_indices],
    'document_frequency': document_frequency[top_df_indices]
})
top_df

,term,document_frequency
0,the,27893
1,and,27423
2,to,26689
3,of,26031
4,a,25905
5,in,25224
6,for,23651
7,is,22739
8,with,21405
9,on,20262


In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer

tfidf_transformer = TfidfTransformer()
X_tfidf = tfidf_transformer.fit_transform(X_count)
idf = tfidf_transformer.idf_
top_idf_indices = np.argsort(idf)[::-1][:20]

top_idf = pd.DataFrame({
    'term': vocabulary[top_idf_indices],
    'idf': idf[top_idf_indices]
})
top_idf

,term,idf
0,00000,10.615839
1,𐌼𐌿𐌽𐌳𐍃,10.615839
2,fluid,10.615839
3,floors,10.615839
4,flexibility,10.615839
5,fixes,10.615839
6,fit,10.615839
7,000000,10.615839
8,000040,10.615839
9,00005,10.615839


In [ ]:
selected_doc_index = 1321
selected_tfidf = X_tfidf[selected_doc_index].toarray().ravel()
top_tfidf_indices = np.argsort(selected_tfidf)[::-1][:20]

top_tfidf = pd.DataFrame({
    'term': vocabulary[top_tfidf_indices],
    'tfidf': selected_tfidf[top_tfidf_indices]
})
top_tfidf

,term,tfidf
0,tracks,0.441718
1,trains,0.293228
2,chicago,0.185874
3,the,0.178509
4,flames,0.158979
5,cold,0.157584
6,warming,0.149404
7,freeze,0.145436
8,passing,0.121448
9,degrees,0.121254


In [ ]:
df1[1321]  # Print the original text of the selected document to verify the content

'So much for Global Warming. We really need to prepare for what happens if we continue down this path of colder winters. The ground freezes and it can freeze even down two-feet. That means good luck on growing anything. It has been so cold, that even the rails for trains broke in three different places between Baltimore and Washington on Thursday. So even trains were delayed or canceled as Amtrak did for trains passing through Chicago.\nIt has been so cold, the tracks freeze and the only way to switch tracks is to set them on fire. Of course, the tracks themselves were not truly burning. They are made out of steel. The fires in Chicago on the tracks this week show flames smoldering in patches of melted snow around the tracks. The photos look wild and dangerous as if it were some apocalyptic image. The steel does not actually burn of melt. The flames simply keep the ice off the tracks at junctions where they need to switch tracks for trains passing through.\nIt is time to start SERIOUS 

- Nhìn vào document 1321, có thể thấy "the" là một từ có df cao nhất trong corpus, nhưng vẫn có chỉ số tfidf xếp thứ 4 trong document 1321 vì có tf đủ lớn. Như vậy, một term có df cao vẫn có thể có tfidf cao.
- Một term có idf cao không nhất thiết có tf-idf cao trong mọi document vì nó còn phụ vào tf của term đó trong document đó vì công thức tfidf phụ thuộc vào cả tf lẫn idf. Trường hợp tf = 0 (không xuất hiện trong doc) thì dù có idf cao nhưng vẫn có tfidf bằng 0.

# PART F

In [ ]:
# Giải phóng các ma trận và danh sách token của Part D để dành RAM cho các pipeline
del df1_tokens, X_count, X_tfidf, count_vectorizer, tfidf_transformer
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer, ENGLISH_STOP_WORDS
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.pre_tokenizers import Whitespace
from tokenizers.trainers import BpeTrainer
import unicodedata

split_index = int(len(df1) * 0.8)
train_docs = df1.iloc[:split_index].tolist()
test_docs = df1.iloc[split_index:].tolist()

# Pipeline A: lowercase + tokenize, giữ punctuation
def tokenize_pipeline_A(text):
    return word_tokenize(text.lower(), preserve_line=True)

# Pipeline B: lowercase + chuẩn hóa punctuation + bỏ stopwords
def tokenize_pipeline_B(text):
    text = unicodedata.normalize('NFKC', text).lower()
    text = re.sub(r"[^\w\s]", " ", text)
    tokens = re.findall(r"\b[a-zA-Z]+\b", text)
    return [token for token in tokens if token not in ENGLISH_STOP_WORDS]

# Huấn luyện BPE trên train split để đo OOV trên held-out split
def train_subword_tokenizer(documents, vocabulary_size=30000):
    subword_model = Tokenizer(BPE(unk_token='[UNK]'))
    subword_model.pre_tokenizer = Whitespace()
    trainer = BpeTrainer(vocab_size=vocabulary_size, min_frequency=2, special_tokens=['[UNK]'])
    normalized_documents = (unicodedata.normalize('NFKC', text).lower() for text in documents)
    subword_model.train_from_iterator(normalized_documents, trainer=trainer, length=len(documents))
    return subword_model

def make_subword_tokenizer(subword_model):
    def tokenize(text):
        text = unicodedata.normalize('NFKC', text).lower()
        return subword_model.encode(text).tokens
    return tokenize

subword_model_C = train_subword_tokenizer(train_docs)
tokenize_pipeline_C = make_subword_tokenizer(subword_model_C)

print('A:', tokenize_pipeline_A(df1.iloc[0])[:20])
print('B:', tokenize_pipeline_B(df1.iloc[0])[:20])
print('C (BPE):', tokenize_pipeline_C(df1.iloc[0])[:20])

A: ['beginners', 'bbq', 'class', 'taking', 'place', 'in', 'missoula', '!', 'do', 'you', 'want', 'to', 'get', 'better', 'at', 'making', 'delicious', 'bbq', '?', 'you']
B: ['beginners', 'bbq', 'class', 'taking', 'place', 'missoula', 'want', 'better', 'making', 'delicious', 'bbq', 'opportunity', 'calendar', 'thursday', 'september', 'join', 'world', 'class', 'bbq', 'champion']
C (BPE): ['beginners', 'bbq', 'class', 'taking', 'place', 'in', 'miss', 'ou', 'la', '!', 'do', 'you', 'want', 'to', 'get', 'better', 'at', 'making', 'delicious', 'bbq']


In [ ]:
queries = [
    'medical image classification',
    'transformer language model',
    'deep learning healthcare',
    'natural language processing',
    'heart attack treatment',
    'myocardial infarction therapy',
    'climate change policy',
    'healthy cells and cancer'
]

def heldout_oov_rate(documents, tokenizer, vocabulary):
    token_count = 0
    oov_count = 0
    for document in documents:
        tokens = tokenizer(document)
        token_count += len(tokens)
        oov_count += sum(token == '[UNK]' or token not in vocabulary for token in tokens)
    return oov_count / token_count if token_count else 0

def evaluate_pipeline(name, tokenizer):
    vectorizer = CountVectorizer(tokenizer=tokenizer, token_pattern=None, lowercase=False)
    X_train = vectorizer.fit_transform(train_docs)
    vocabulary = set(vectorizer.vocabulary_)
    query_tokens = [token for query in queries for token in tokenizer(query)]
    query_oov = sum(token == '[UNK]' or token not in vocabulary for token in query_tokens)
    n_documents, n_terms = X_train.shape
    return {
        'Pipeline': name,
        'Vocabulary size': n_terms,
        'Average tokens/document': X_train.sum() / n_documents,
        'Matrix sparsity': 1 - X_train.nnz / (n_documents * n_terms),
        'OOV rate': heldout_oov_rate(test_docs, tokenizer, vocabulary),
        'Query OOV rate': query_oov / len(query_tokens) if query_tokens else 0
    }

metrics = pd.DataFrame([
    evaluate_pipeline('Pipeline A', tokenize_pipeline_A),
    evaluate_pipeline('Pipeline B', tokenize_pipeline_B),
    evaluate_pipeline('Pipeline C', tokenize_pipeline_C)
])
metrics

,Pipeline,Vocabulary size,Average tokens/document,Matrix sparsity,OOV rate,Query OOV rate
0,Pipeline A,263260,401.599375,0.999309,0.024311,0.0
1,Pipeline B,148189,190.004333,0.999189,0.029680,0.0
2,Pipeline C,29667,455.142542,0.993599,0.000240,0.0


# PART G — Tìm kiếm tài liệu

Dùng các query gợi ý trong đề; mỗi pipeline có vocabulary và TF-IDF index riêng. ID tài liệu là chỉ số dòng corpus bắt đầu từ 0.

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

if not 5 <= len(queries) <= 10:
    raise ValueError('Khai báo từ 5 đến 10 query trong cell này.')
lab_dir = Path('lab1') if Path('lab1/experiments.ipynb').is_file() else Path('.')
RESULTS_PATH = lab_dir / 'results.csv'
pipelines = {'Pipeline A': tokenize_pipeline_A, 'Pipeline B': tokenize_pipeline_B, 'Pipeline C': tokenize_pipeline_C}

result_rows = []
for pipeline_name, tokenizer in pipelines.items():
    search_bpe = None
    if pipeline_name == 'Pipeline C':
        search_bpe = train_subword_tokenizer(df1.tolist())
        tokenizer = make_subword_tokenizer(search_bpe)
    vectorizer = CountVectorizer(tokenizer=tokenizer, token_pattern=None, lowercase=False)
    X_count_search = vectorizer.fit_transform(df1)
    tfidf_transformer = TfidfTransformer()
    X_docs_tfidf = tfidf_transformer.fit_transform(X_count_search)
    X_queries_tfidf = tfidf_transformer.transform(vectorizer.transform(queries))
    scores = cosine_similarity(X_queries_tfidf, X_docs_tfidf)
    terms = vectorizer.get_feature_names_out()
    for query_index, query in enumerate(queries):
        query_vector = X_queries_tfidf.getrow(query_index)
        top_indices = scores[query_index].argsort()[-5:][::-1]
        for rank, document_id in enumerate(top_indices, start=1):
            contributions = query_vector.multiply(X_docs_tfidf.getrow(document_id)).tocoo()
            contribution_order = np.argsort(contributions.data)[::-1][:5]
            top_terms = '; '.join(
                f"{terms[contributions.col[index]]}:{contributions.data[index]:.4f}"
                for index in contribution_order
            )
            result_rows.append({
                'pipeline': pipeline_name, 'query': query, 'rank': rank,
                'document_id': int(document_id),
                'similarity': float(scores[query_index, document_id]),
                'document_preview': str(df1.iloc[document_id])[:250].replace('\n', ' '),
                'top_contributing_terms': top_terms,
                'relevant': ''
            })
    del X_count_search, X_docs_tfidf, X_queries_tfidf, scores, vectorizer, tfidf_transformer
    if search_bpe is not None:
        del search_bpe

results = pd.DataFrame(result_rows)
results = results.merge(metrics, left_on='pipeline', right_on='Pipeline').drop(columns='Pipeline')
results.to_csv(RESULTS_PATH, index=False)
candidate_pool = results[['query', 'document_id', 'document_preview']].drop_duplicates()
candidate_pool.sort_values(['query', 'document_id'])

,query,document_id,document_preview
31,climate change policy,2231,The Rupa Lake Cooperative in Nepal. Credit: Bi...
33,climate change policy,9209,Climate Change Denial: Why It can Be Hugely In...
114,climate change policy,10454,The potential impact of global climate change ...
34,climate change policy,11682,The existential threats to people and the envi...
30,climate change policy,19142,Reduce short-lived climate change or even this...
32,climate change policy,20926,Big polluting corporations like Shell and Chev...
74,climate change policy,21770,See this article detailing that arctic sea ice...
51,deep learning healthcare,6123,"With today’s advancement in technology, it is ..."
13,deep learning healthcare,7564,"would I have learned, learnt? would you have l..."
14,deep learning healthcare,9229,"Influence Health, the healthcare industry’s le..."


# PART H — Đánh giá

In [42]:
from pathlib import Path
import numpy as np
import pandas as pd

result_candidates = [Path('lab1/results.csv'), Path('results.csv')]
results_path = next((path for path in result_candidates if path.is_file()), None)
if results_path is None:
    raise FileNotFoundError('Không tìm thấy results.csv trong lab1 hoặc working directory.')
reviewed_results = pd.read_csv(results_path).fillna({'relevant': ''})
raw_labels = reviewed_results['relevant'].astype(str).str.strip()
if not raw_labels.isin(['', '0', '1', '0.0', '1.0']).all():
    raise ValueError('Cột relevant chỉ nhận 1 (liên quan), 0 (không liên quan), hoặc để trống dòng trùng.')
reviewed_results['relevant'] = pd.to_numeric(raw_labels.mask(raw_labels.eq(''), pd.NA), errors='coerce').astype('Int64')
label_sets = reviewed_results.groupby(['query', 'document_id'])['relevant'].agg(lambda values: set(values.dropna().astype(int)))
missing_labels = [key for key, values in label_sets.items() if not values]
conflicting_labels = [key for key, values in label_sets.items() if len(values) > 1]
if missing_labels:
    raise ValueError(f"Chưa gán nhãn cho {len(missing_labels)} cặp query/document trong candidate pool.")
if conflicting_labels:
    raise ValueError('Các dòng trùng query/document phải có nhãn relevance nhất quán giữa các pipeline.')
label_map = {key: next(iter(values)) for key, values in label_sets.items()}
reviewed_results['relevant'] = [label_map[(query, int(document_id))] for query, document_id in zip(reviewed_results['query'], reviewed_results['document_id'])]
judgments = reviewed_results[['query', 'document_id', 'relevant']].drop_duplicates()
relevant_by_query = {
    query: set(group.loc[group['relevant'] == 1, 'document_id'].astype(int))
    for query, group in judgments.groupby('query')
}
all_queries = reviewed_results['query'].drop_duplicates().tolist()
excluded_queries = [query for query in all_queries if not relevant_by_query.get(query)]
print('Query không có relevant document trong candidate pool; loại khỏi trung bình metric:', excluded_queries)

metric_rows = []
for pipeline_name, group in reviewed_results.groupby('pipeline'):
    precisions, recalls, reciprocal_ranks = [], [], []
    for query in all_queries:
        relevant_ids = relevant_by_query.get(query, set())
        if not relevant_ids:
            continue
        ranked_ids = group[group['query'] == query].sort_values('rank')['document_id'].astype(int).tolist()
        hits = [document_id in relevant_ids for document_id in ranked_ids[:5]]
        precisions.append(sum(hits) / 5)
        recalls.append(sum(hits) / len(relevant_ids))
        first_hit_rank = next((rank for rank, hit in enumerate(hits, start=1) if hit), None)
        reciprocal_ranks.append(1 / first_hit_rank if first_hit_rank else 0)
    metric_rows.append({
        'pipeline': pipeline_name,
        'Precision@5': np.mean(precisions) if precisions else np.nan,
        'Recall@5': np.mean(recalls) if recalls else np.nan,
        'MRR': np.mean(reciprocal_ranks) if reciprocal_ranks else np.nan
    })

evaluation_metrics = pd.DataFrame(metric_rows, columns=['pipeline', 'Precision@5', 'Recall@5', 'MRR'])
preprocessing_columns = ['Vocabulary size', 'Average tokens/document', 'Matrix sparsity', 'OOV rate', 'Query OOV rate']
preprocessing_metrics = reviewed_results.groupby('pipeline', as_index=False)[preprocessing_columns].first()
metrics = preprocessing_metrics.merge(evaluation_metrics, on='pipeline')
reviewed_results = reviewed_results.merge(evaluation_metrics, on='pipeline', how='left')
reviewed_results.to_csv(results_path, index=False)
metrics

Query không có relevant document trong candidate pool; loại khỏi trung bình metric: ['medical image classification', 'transformer language model', 'deep learning healthcare', 'natural language processing']


,pipeline,Vocabulary size,Average tokens/document,Matrix sparsity,OOV rate,Query OOV rate,Precision@5,Recall@5,MRR
0,Pipeline A,263260,401.599375,0.999309,0.024311,0.0,0.60,0.928571,0.875000
1,Pipeline B,148189,190.004333,0.999189,0.029680,0.0,0.60,0.928571,0.833333
2,Pipeline C,29667,455.142542,0.993599,0.000240,0.0,0.55,0.878571,0.875000


# PART I — Phân tích lỗi
## 1. Hai query có kết quả tốt
### 1.1 "climate change policy" — P@5 = 1.00, MRR = 1.00 (cả 3 pipeline)

Cả 5 documents trả về ở mỗi pipeline đều liên quan đến biến đổi khí hậu và chính sách.
- Document 19142 (sim ≈ 0.60–0.65) chứa nhiều lần lặp lại trực tiếp cả ba token "climate", "change" và "policy" (hoặc các từ đồng xuất hiện như "government", "greenhouse effect"). Điều này khiến cả TF lẫn cosine similarity đều cao.
- `climate` đóng góp lớn nhất (≈ 0.44–0.46), vì IDF cao — "climate" không phải từ thông dụng trong corpus 30K documents.
- `change` đóng góp thứ hai (≈ 0.10–0.19). IDF thấp hơn vì "change" xuất hiện trong nhiều ngữ cảnh khác (đổi tiền, thay đổi cài đặt...).
- `policy` đóng góp nhỏ hơn (≈ 0.04–0.08) vì không phải document nào về climate change cũng dùng đúng từ "policy".
- Cả query lẫn documents đều chia sẻ trực tiếp các token "climate", "change", "policy" — lexical overlap hoàn toàn.
- Query "climate change policy" dùng đúng thuật ngữ phổ biến nhất mà các bài viết về chủ đề này sử dụng. Không có synonymy hay paraphrase; hầu hết tài liệu về biến đổi khí hậu đều dùng cụm từ "climate change" nguyên bản.

### 1.2 "healthy cells and cancer" — P@5 = 1.00 (Pipeline A, B), 0.80 (Pipeline C), MRR = 1.00
- Document 29938 (sim ≈ 0.49–0.53) mô tả thiết bị thu thập "cancer cells" từ máu. Hai token "cancer" và "cells" đều có IDF cao trong corpus tổng quát vì chúng là thuật ngữ chuyên ngành y khoa.
- `cells` (≈ 0.25–0.29) và `cancer` (≈ 0.22–0.24) chiếm phần lớn similarity. Cả hai đều có IDF cao vì ít xuất hiện ngoài ngữ cảnh y sinh.
- `healthy` đóng góp nhỏ (≈ 0.03) vì không phải document nào về cancer cũng nói về "healthy cells".
- `and` đóng góp gần như bằng 0 vì IDF rất thấp (stopword).
- Document 7148 ("Checking for Cancer Lacrosse Invitational") — một giải thể thao từ thiện — lọt vào Top-5 vì chứa từ "cancer" nhiều lần. Đây là false positive do lexical matching không phân biệt được nghĩa y khoa và nghĩa tên sự kiện.

## 2. Hai query có kết quả kém
### 2.1 "transformer language model" — P@5 = 0.00, MRR = 0.00 (cả 3 pipeline)
Document 27936 (sim ≈ 0.28–0.50) nói về "transformer / circuit board on my Hobby 720 uml model 2003" — một biến thế điện. TF-IDF xếp hạng nó cao nhất vì:
- `transformer`: IDF rất cao (hiếm trong corpus) → đóng góp lớn (≈ 0.42–0.45).
- `model`: đóng góp nhỏ hơn (≈ 0.06) vì IDF thấp.
- Có overlap về mặt token (`transformer`, `model`) nhưng hoàn toàn khác nghĩa. "Transformer" trong query ám chỉ kiến trúc neural network (Vaswani et al., 2017), còn trong document là máy biến thế điện.
- Có thể tồn tại documents về "Transformer architecture" hoặc "language model" trong corpus, nhưng chúng có thể dùng cụm từ khác (ví dụ "attention mechanism", "BERT", "GPT") nên không match.
- **Polysemy (đa nghĩa):** "transformer" có nhiều nghĩa; TF-IDF không phân biệt được nghĩa.
- **Lexical matching không hiểu ngữ cảnh:** Hệ thống so khớp từng token riêng lẻ, không xét ngữ cảnh "transformer" đi cùng "language model".
- Các documents xếp hạng 2–5 đều về "language" nhưng trong nghĩa ngôn ngữ tự nhiên (tiếng Pháp, tiếng Tây Ban Nha, cài đặt Facebook), không phải NLP.

### 2.2 "natural language processing" — P@5 = 0.00, MRR = 0.00 (cả 3 pipeline)
- Document 25428 (sim ≈ 0.35–0.39) hướng dẫn đổi ngôn ngữ Facebook trên iPhone. Token `language` có IDF vừa phải, đủ để kéo similarity lên mặc dù nội dung không liên quan.
- `language` chiếm gần toàn bộ similarity (≈ 0.27–0.39) vì "natural" và "processing" đóng góp rất khác nhau:
  - `processing` chỉ match với document 8705 (quy định "Food Safety and Standards" — "food processing"), hoàn toàn lệch nghĩa.
  - `natural` match với document 5699 ("Truth-value gaps in natural language") — liên quan đến ngôn ngữ tự nhiên nhưng là luận văn triết học ngôn ngữ, không phải NLP.
- **Polysemy:** "language" (ngôn ngữ con người vs. lĩnh vực NLP), "processing" (chế biến thực phẩm vs. xử lý dữ liệu), "natural" (tự nhiên vs. "natural language" như cụm từ).
- **Thiếu understanding về multi-word expression:** TF-IDF tách "natural language processing" thành ba unigram độc lập, mất đi nghĩa của cụm từ ghép.

## 3. Failure case quan trọng nhất: "heart attack treatment" vs "myocardial infarction therapy"
### Bối cảnh
Hai query này **hoàn toàn đồng nghĩa về mặt y khoa:**
- "heart attack" = "myocardial infarction" (nhồi máu cơ tim)
- "treatment" = "therapy" (điều trị)
Tuy nhiên, **không có bất kỳ token nào trùng nhau** giữa hai query. Đây là hiện tượng (đồng nghĩa).

### Phân tích chi tiết
- Document 6173 (therapy cho nghiện) có similarity ≈ 0.22–0.25 vì token `therapy` có IDF cao (chuyên ngành) và xuất hiện nhiều lần trong document. Trong khi đó, `myocardial` và `infarction` là thuật ngữ cực kỳ hiếm (IDF rất cao) nhưng cũng cực kỳ ít xuất hiện trong corpus, nên dù IDF cao, số document chứa chúng rất ít.
- Query "heart attack treatment" tìm được doc 799 (về heart attack) nhưng **không** tìm được doc 26337 (về myocardial infarction) — dù cùng chủ đề. Ngược lại, query "myocardial infarction therapy" tìm được doc 26337 nhưng **không** tìm được doc 799. Hai tập kết quả gần như rời nhau hoàn toàn, mặc dù cùng mô tả một bệnh lý.
- Điều này chứng minh TF-IDF cosine similarity **chỉ nắm bắt lexical overlap**, không có khả năng hiểu rằng hai biểu thức khác nhau có thể mang cùng một nghĩa.



# PART J — Từ failure đến representation tiếp theo
Bước tiếp theo trong NLP là chuyển từ **biểu diễn thưa dựa trên từ vựng** (sparse lexical representation) sang **biểu diễn dày dựa trên phân bố** (dense distributional representation), nơi ý nghĩa của từ được mã hóa bởi ngữ cảnh sử dụng thay vì chỉ bởi sự hiện diện hay vắng mặt của token.